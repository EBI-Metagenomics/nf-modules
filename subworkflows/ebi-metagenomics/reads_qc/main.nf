
include { SEQFU_CHECK            } from '../../../modules/ebi-metagenomics/seqfu/check/main'
include { PIMENTO_GENERATEBCV    } from '../../../modules/ebi-metagenomics/pimento/generatebcv/main'
include { LIBRARYSTRATEGYCHECK    } from '../../../modules/ebi-metagenomics/librarystrategycheck/main'
include { FASTQSUFFIXHEADERCHECK } from '../../../modules/ebi-metagenomics/fastqsuffixheadercheck/main'
include { FASTP                  } from '../../../modules/ebi-metagenomics/fastp/main'
include { SEQTK_SEQ              } from '../../../modules/ebi-metagenomics/seqtk/seq/main'

workflow  READS_QC {

    take:
    filter_amplicon // channel: val(boolean)
    ch_reads    // channel: [ val(meta), [ fastq ] ]
    save_merged // channel:  val(boolean)

    main:
    ch_versions = Channel.empty()

    SEQFU_CHECK(ch_reads)
    ch_versions = ch_versions.mix(SEQFU_CHECK.out.versions.first())

    passed_seqfu_reads = SEQFU_CHECK.out.tsv
        .splitCsv(sep: "\t", elem: 1)
        .filter { meta, seqfu_res ->
            seqfu_res[0] == "OK"
        }
        .map { map, seqfu_res -> map }
        .join(ch_reads)

    FASTQSUFFIXHEADERCHECK(passed_seqfu_reads)
    ch_versions = ch_versions.mix(FASTQSUFFIXHEADERCHECK.out.versions.first())

    passed_suffixheader_reads = FASTQSUFFIXHEADERCHECK.out.json
        .filter { meta, sufhd_res ->
            sufhd_res.countLines() == 0
        }
        .map { meta, _ -> [ meta ] }
        .join(ch_reads)

    if ( filter_amplicon ) {
        generatebcv_input = passed_suffixheader_reads
                        .map { meta, fastq ->
                            if ( meta.single_end ) {
                                [ meta, "auto", "auto", fastq ]
                            } else {
                                [ meta, "auto", "auto", fastq[0] ]
                            }
                        }

        PIMENTO_GENERATEBCV(generatebcv_input)
        ch_versions = ch_versions.mix(PIMENTO_GENERATEBCV.out.versions.first())

        LIBRARYSTRATEGYCHECK(PIMENTO_GENERATEBCV.out.tsv)
        ch_versions = ch_versions.mix(LIBRARYSTRATEGYCHECK.out.versions.first())

        fastp_input = LIBRARYSTRATEGYCHECK.out.library_check_out
                        .filter { meta, strategy ->
                            strategy == "AMPLICON"
                        }
                        .map { meta, _ -> [ meta ] }
                        .join(ch_reads)

        amplicon_check = LIBRARYSTRATEGYCHECK.out.library_check_out
    } else {
        fastp_input = passed_suffixheader_reads
        amplicon_check = Channel.empty()
    }

    FASTP ( fastp_input, params.save_trimmed_fail, save_merged )
    ch_versions = ch_versions.mix(FASTP.out.versions.first())

    ch_se_fastp_reads = FASTP
        .out.reads
        .filter { it[0].single_end }

    ch_reads_se_and_merged = ch_se_fastp_reads
        .mix(FASTP.out.reads_merged)

    SEQTK_SEQ(ch_reads_se_and_merged)
    ch_versions = ch_versions.mix(SEQTK_SEQ.out.versions.first())

    emit:
    seqfu_check           = SEQFU_CHECK.out.tsv                        // channel: [ val(meta), tsv  ]
    suffix_header_check   = FASTQSUFFIXHEADERCHECK.out.json            // channel: [ val(meta), json  ]
    amplicon_check        = amplicon_check                             // channel: [ val(meta), env  ]
    reads                 = FASTP.out.reads                            // channel: [ val(meta), [ fastq ] ]
    reads_se_and_merged   = ch_reads_se_and_merged                     // channel: [ val(meta), [ fastq ] ]
    fastp_summary_json    = FASTP.out.json                             // channel: [ val(meta), [ json ] ]
    reads_fasta           = SEQTK_SEQ.out.fastx                        // channel: [ val(meta), [ fasta ] ]
    versions              = ch_versions                                // channel: [ versions.yml ]
}

