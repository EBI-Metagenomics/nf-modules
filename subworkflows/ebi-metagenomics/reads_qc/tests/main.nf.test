nextflow_workflow {

    name "Test Subworkflow READS_QC"
    script "../main.nf"
    workflow "READS_QC"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/reads_qc"
    tag "seqfu/check"
    tag "fastqsuffixheadercheck"
    tag "assessmcpproportions"
    tag "fastp"
    tag "seqtk"
    tag "seqtk/seq"

    test("test_reads_qc_success") {

        when {
            params {
                save_trimmed_fail = true
            }
            workflow {
                """
                input[0] = true
                input[1] = Channel.fromList([
                        [
                            [ id:'SRR21814853', single_end:false ], // meta map
                            [ file("${moduleDir}/tests/fixtures/SRR21814853_1.fastq.gz", checkIfExists: true),
                            file("${moduleDir}/tests/fixtures/SRR21814853_2.fastq.gz", checkIfExists: true)
                            ]
                        ]
                    ])
                input[2] = true
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.seqfu_check,
                    workflow.out.suffix_header_check,
                    workflow.out.amplicon_check,
                    workflow.out.reads,
                    workflow.out.reads_se_and_merged,
                    workflow.out.fastp_summary_json,
                    workflow.out.reads_fasta,
                    workflow.out.versions.collect { path(it).yaml }
                    ).match()
                }
            )
        }
    }

    test("test_reads_fail_seqfu") {

        when {
            params {
                save_trimmed_fail = true
            }
            workflow {
                """
                input[0] = true
                input[1] = Channel.fromList([
                    [
                        [ id:'SRR21814853', single_end:false ], // meta map
                        [ file("${moduleDir}/tests/fixtures/SRR21814853_1.fastq.gz", checkIfExists: true),
                        file("${moduleDir}/tests/fixtures/SRR21814853_2.fastq.gz", checkIfExists: true)
                        ]
                    ],
                    [
                        [ id:'ERR4674038', single_end:false ], // meta map
                        [ file("${moduleDir}/tests/fixtures/ERR4674038_test_readcount_1.fastq", checkIfExists: true),
                        file("${moduleDir}/tests/fixtures/ERR4674038_test_readcount_2.fastq", checkIfExists: true)
                        ]
                    ]
                ])
                input[2] = true
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.seqfu_check,
                    workflow.out.suffix_header_check,
                    workflow.out.amplicon_check,
                    workflow.out.reads,
                    workflow.out.reads_se_and_merged,
                    workflow.out.fastp_summary_json,
                    workflow.out.reads_fasta,
                    workflow.out.versions.collect { path(it).yaml }
                    ).match()
                }
            )
        }
    }

    test("test_reads_fail_suffixheader") {

        when {
            params {
                save_trimmed_fail = true
            }
            workflow {
                """
                input[0] = true
                input[1] = Channel.fromList([
                    [
                        [ id:'SRR21814853', single_end:false ], // meta map
                        [ file("${moduleDir}/tests/fixtures/SRR21814853_1.fastq.gz", checkIfExists: true),
                        file("${moduleDir}/tests/fixtures/SRR21814853_2.fastq.gz", checkIfExists: true)
                        ]
                    ],
                    [
                        [ id:'ERR4674038', single_end:false ], // meta map
                        [ file("${moduleDir}/tests/fixtures/ERR4674038witherror_1.fastq.gz", checkIfExists: true),
                        file("${moduleDir}/tests/fixtures/ERR4674038witherror_2.fastq.gz", checkIfExists: true)
                        ]
                    ]
                ])
                input[2] = true
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.seqfu_check,
                    workflow.out.suffix_header_check,
                    workflow.out.amplicon_check,
                    workflow.out.reads,
                    workflow.out.reads_se_and_merged,
                    workflow.out.fastp_summary_json,
                    workflow.out.reads_fasta,
                    workflow.out.versions.collect { path(it).yaml }
                    ).match()
                }
            )
        }
    }

    test("test_reads_onlyfail") {

        when {
            params {
                save_trimmed_fail = true
            }
            workflow {
                """
                input[0] = true
                input[1] = Channel.fromList([
                    [
                        [ id:'ERR4674038', single_end:false ], // meta map
                        [ file("${moduleDir}/tests/fixtures/ERR4674038witherror_1.fastq.gz", checkIfExists: true),
                        file("${moduleDir}/tests/fixtures/ERR4674038witherror_2.fastq.gz", checkIfExists: true)
                        ]
                    ]
                ])
                input[2] = true
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.seqfu_check,
                    workflow.out.suffix_header_check,
                    workflow.out.amplicon_check,
                    workflow.out.reads,
                    workflow.out.reads_se_and_merged,
                    workflow.out.fastp_summary_json,
                    workflow.out.reads_fasta,
                    workflow.out.versions.collect { path(it).yaml }
                    ).match()
                }
            )
        }
    }

    test("test_reads_qc_notamplicon") {

        when {
            params {
                save_trimmed_fail = true
            }
            workflow {
                """
                input[0] = false
                input[1] = Channel.fromList([
                        [
                            [ id:'SRR21814853', single_end:false ], // meta map
                            [ file("${moduleDir}/tests/fixtures/SRR21814853_1.fastq.gz", checkIfExists: true),
                            file("${moduleDir}/tests/fixtures/SRR21814853_2.fastq.gz", checkIfExists: true)
                            ]
                        ]
                    ])
                input[2] = true
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.seqfu_check,
                    workflow.out.suffix_header_check,
                    workflow.out.amplicon_check,
                    workflow.out.reads,
                    workflow.out.reads_se_and_merged,
                    workflow.out.fastp_summary_json,
                    workflow.out.reads_fasta,
                    workflow.out.versions.collect { path(it).yaml }
                    ).match()
                }
            )
        }
    }
}
