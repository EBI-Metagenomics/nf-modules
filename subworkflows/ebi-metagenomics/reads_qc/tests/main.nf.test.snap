{
    "test_reads_fail_seqfu": {
        "content": [
            [
                [
                    {
                        "id": "ERR4674038",
                        "single_end": false
                    },
                    "ERR4674038.tsv:md5,3aa17b2bca2980dbaad41ba5e54c0b44"
                ],
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853.tsv:md5,008673dad4dd80a742cc11669f24d6e8"
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853_suffix_header_err.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "AMPLICON"
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    [
                        "SRR21814853_1.fastp.fastq.gz:md5,ab81a64d069ed8876432d8e3bad0b847",
                        "SRR21814853_2.fastp.fastq.gz:md5,f8bf866d116205baa454b7a501a67b62"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853.merged.fastq.gz:md5,58ededca56f023aa625f51ec241ba849"
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853.fastp.json:md5,f2fa2740350c702e0f7d6adec47858c4"
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853.seqtk-seq.fasta.gz:md5,a14b76b1e8c24d3f9d98e61d03a62a53"
                ]
            ],
            [
                {
                    "READS_QC:LIBRARYSTRATEGYCHECK": {
                        "mgnify-pipelines-toolkit": "1.2.1"
                    }
                },
                {
                    "READS_QC:PIMENTO_GENERATEBCV": {
                        "mi-pimento": "1.0.2"
                    }
                },
                {
                    "READS_QC:SEQFU_CHECK": {
                        "seqfu": "1.20.3"
                    }
                },
                {
                    "READS_QC:SEQTK_SEQ": {
                        "seqtk": "1.3-r106"
                    }
                },
                {
                    "READS_QC:FASTQSUFFIXHEADERCHECK": {
                        "mgnify-pipelines-toolkit": "1.2.11"
                    }
                },
                {
                    "READS_QC:FASTP": {
                        "fastp": "0.23.4"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-06T12:19:42.502928"
    },
    "test_reads_onlyfail": {
        "content": [
            [
                [
                    {
                        "id": "ERR4674038",
                        "single_end": false
                    },
                    "ERR4674038.tsv:md5,e56475d0fee9cdc4fb3eda621f924af9"
                ]
            ],
            [
                [
                    {
                        "id": "ERR4674038",
                        "single_end": false
                    },
                    "ERR4674038_suffix_header_err.json:md5,735222eb9c22d4d81b312a9682d55679"
                ]
            ],
            [
                
            ],
            [
                
            ],
            [
                
            ],
            [
                
            ],
            [
                
            ],
            [
                {
                    "READS_QC:SEQFU_CHECK": {
                        "seqfu": "1.20.3"
                    }
                },
                {
                    "READS_QC:FASTQSUFFIXHEADERCHECK": {
                        "mgnify-pipelines-toolkit": "1.2.11"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-06T12:20:24.053887"
    },
    "test_reads_qc_success": {
        "content": [
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853.tsv:md5,008673dad4dd80a742cc11669f24d6e8"
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853_suffix_header_err.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "AMPLICON"
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    [
                        "SRR21814853_1.fastp.fastq.gz:md5,ab81a64d069ed8876432d8e3bad0b847",
                        "SRR21814853_2.fastp.fastq.gz:md5,f8bf866d116205baa454b7a501a67b62"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853.merged.fastq.gz:md5,58ededca56f023aa625f51ec241ba849"
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853.fastp.json:md5,f2fa2740350c702e0f7d6adec47858c4"
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853.seqtk-seq.fasta.gz:md5,a14b76b1e8c24d3f9d98e61d03a62a53"
                ]
            ],
            [
                {
                    "READS_QC:LIBRARYSTRATEGYCHECK": {
                        "mgnify-pipelines-toolkit": "1.2.1"
                    }
                },
                {
                    "READS_QC:PIMENTO_GENERATEBCV": {
                        "mi-pimento": "1.0.2"
                    }
                },
                {
                    "READS_QC:SEQFU_CHECK": {
                        "seqfu": "1.20.3"
                    }
                },
                {
                    "READS_QC:SEQTK_SEQ": {
                        "seqtk": "1.3-r106"
                    }
                },
                {
                    "READS_QC:FASTQSUFFIXHEADERCHECK": {
                        "mgnify-pipelines-toolkit": "1.2.11"
                    }
                },
                {
                    "READS_QC:FASTP": {
                        "fastp": "0.23.4"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-06T12:19:12.386777"
    },
    "test_reads_fail_suffixheader": {
        "content": [
            [
                [
                    {
                        "id": "ERR4674038",
                        "single_end": false
                    },
                    "ERR4674038.tsv:md5,e56475d0fee9cdc4fb3eda621f924af9"
                ],
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853.tsv:md5,008673dad4dd80a742cc11669f24d6e8"
                ]
            ],
            [
                [
                    {
                        "id": "ERR4674038",
                        "single_end": false
                    },
                    "ERR4674038_suffix_header_err.json:md5,735222eb9c22d4d81b312a9682d55679"
                ],
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853_suffix_header_err.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "AMPLICON"
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    [
                        "SRR21814853_1.fastp.fastq.gz:md5,ab81a64d069ed8876432d8e3bad0b847",
                        "SRR21814853_2.fastp.fastq.gz:md5,f8bf866d116205baa454b7a501a67b62"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853.merged.fastq.gz:md5,58ededca56f023aa625f51ec241ba849"
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853.fastp.json:md5,f2fa2740350c702e0f7d6adec47858c4"
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853.seqtk-seq.fasta.gz:md5,a14b76b1e8c24d3f9d98e61d03a62a53"
                ]
            ],
            [
                {
                    "READS_QC:LIBRARYSTRATEGYCHECK": {
                        "mgnify-pipelines-toolkit": "1.2.1"
                    }
                },
                {
                    "READS_QC:PIMENTO_GENERATEBCV": {
                        "mi-pimento": "1.0.2"
                    }
                },
                {
                    "READS_QC:SEQFU_CHECK": {
                        "seqfu": "1.20.3"
                    }
                },
                {
                    "READS_QC:SEQTK_SEQ": {
                        "seqtk": "1.3-r106"
                    }
                },
                {
                    "READS_QC:FASTQSUFFIXHEADERCHECK": {
                        "mgnify-pipelines-toolkit": "1.2.11"
                    }
                },
                {
                    "READS_QC:FASTP": {
                        "fastp": "0.23.4"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-06T12:20:11.608794"
    },
    "test_reads_qc_notamplicon": {
        "content": [
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853.tsv:md5,008673dad4dd80a742cc11669f24d6e8"
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853_suffix_header_err.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ],
            [
                
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    [
                        "SRR21814853_1.fastp.fastq.gz:md5,ab81a64d069ed8876432d8e3bad0b847",
                        "SRR21814853_2.fastp.fastq.gz:md5,f8bf866d116205baa454b7a501a67b62"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853.merged.fastq.gz:md5,58ededca56f023aa625f51ec241ba849"
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853.fastp.json:md5,f2fa2740350c702e0f7d6adec47858c4"
                ]
            ],
            [
                [
                    {
                        "id": "SRR21814853",
                        "single_end": false
                    },
                    "SRR21814853.seqtk-seq.fasta.gz:md5,a14b76b1e8c24d3f9d98e61d03a62a53"
                ]
            ],
            [
                {
                    "READS_QC:SEQFU_CHECK": {
                        "seqfu": "1.20.3"
                    }
                },
                {
                    "READS_QC:SEQTK_SEQ": {
                        "seqtk": "1.3-r106"
                    }
                },
                {
                    "READS_QC:FASTQSUFFIXHEADERCHECK": {
                        "mgnify-pipelines-toolkit": "1.2.11"
                    }
                },
                {
                    "READS_QC:FASTP": {
                        "fastp": "0.23.4"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-06T12:20:45.764059"
    }
}